Zend/zend_exceptions.c

```diff
@@ -587,6 +587,39 @@ static void _build_trace_string(smart_str *str, HashTable *ht, uint32_t num) /*
 }
 /* }}} */
 
+ZEND_API zend_string *zend_trace_function_args_to_string(HashTable *frame) {
+	zval *tmp;
+	smart_str str = {0};
+	/* init because ASan will complain */
+	smart_str_appends(&str, "");
+
+	tmp = zend_hash_find_known_hash(frame, ZSTR_KNOWN(ZEND_STR_ARGS));
+	if (tmp) {
+		if (Z_TYPE_P(tmp) == IS_ARRAY) {
+			size_t last_len = ZSTR_LEN(str.s);
+			zend_string *name;
+			zval *arg;
+
+			ZEND_HASH_FOREACH_STR_KEY_VAL(Z_ARRVAL_P(tmp), name, arg) {
+				if (name) {
+					smart_str_append(&str, name);
+					smart_str_appends(&str, ": ");
+				}
+				_build_trace_args(arg, &str);
+			} ZEND_HASH_FOREACH_END();
+
+			if (last_len != ZSTR_LEN(str.s)) {
+				ZSTR_LEN(str.s) -= 2; /* remove last ', ' */
+			}
+		} else {
+			smart_str_appends(&str, "<<invalid argument array>>");
+		}
```

Comment:
Can this ever happen?

Reply:
I copied it from `build_trace_string`, which emits an error, which I didn't want to do considering we'd be in this code path because of an error. It's only used there because it seems userland code can muck around and make it be defensive (`Zend/tests/bug63762.phpt`).

(The common code path could probably be shared though...)

Reply:
It might make sense to extract the common part indeed, maybe put it in exception.c and corresponding header. Would allow to get rid of the newly introduced dependency on built-in functions header.